sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs

```diff
@@ -0,0 +1,233 @@
+        /// end at _lastConsumable.
+        /// </summary>
+        private DateTimeOffset? _endTime;
+
+        public ChangeFeed(
+            BlobContainerClient containerClient,
+            SegmentFactory segmentFactory,
+            Queue<string> years,
+            Queue<string> segments,
+            Segment currentSegment,
+            DateTimeOffset lastConsumable,
+            DateTimeOffset? startTime,
+            DateTimeOffset? endTime)
+        {
+            _containerClient = containerClient;
+            _segmentFactory = segmentFactory;
+            _years = years;
+            _segments = segments;
+            _currentSegment = currentSegment;
+            _lastConsumable = lastConsumable;
+            _startTime = startTime;
+            _endTime = endTime;
+        }
+
+        /// <summary>
+        /// Constructor for mocking, and for creating a Change Feed with no Events.
+        /// </summary>
+        public ChangeFeed() { }
+
+        // The last segment may still be adding chunks.
+        public async Task<Page<BlobChangeFeedEvent>> GetPage(
+            bool async,
+            int pageSize = Constants.ChangeFeed.DefaultPageSize,
+            CancellationToken cancellationToken = default)
+        {
+            if (!HasNext())
+            {
+                throw new InvalidOperationException("Change feed doesn't have any more events");
+            }
+
+            if (_currentSegment.DateTime > _endTime)
+            {
+                return new BlobChangeFeedEventPage();
+            }
+
+            if (_currentSegment.DateTime > _lastConsumable)
+            {
+                return new BlobChangeFeedEventPage();
+            }
+
+            // Get next page
+            List<BlobChangeFeedEvent> blobChangeFeedEvents = new List<BlobChangeFeedEvent>();
+
+            int remainingEvents = pageSize;
+            while (blobChangeFeedEvents.Count < pageSize
+                && HasNext())
+            {
+                List<BlobChangeFeedEvent> newEvents = await _currentSegment.GetPage(
+                    async,
+                    remainingEvents,
+                    cancellationToken).ConfigureAwait(false);
+                blobChangeFeedEvents.AddRange(newEvents);
+                remainingEvents -= newEvents.Count;
+                await AdvanceSegmentIfNecessary(async).ConfigureAwait(false);
+            }
+
+            return new BlobChangeFeedEventPage(blobChangeFeedEvents, JsonSerializer.Serialize<ChangeFeedCursor>(GetCursor()));
+        }
+
+        public bool HasNext()
+        {
+            // We have no more segments, years, and the current segment doesn't have hext.
+            if (_segments.Count == 0 && _years.Count == 0  && !_currentSegment.HasNext())
+            {
+                return false;
+            }
+
+            DateTimeOffset end = BlobChangeFeedExtensions.MinDateTime(_lastConsumable, _endTime);
+
+            return _currentSegment.DateTime <= end;
```

Comment:
why returns true when `_currentSegment.DateTime == end;`

Reply:
Because we want to include the end hour.

Reply:
Actually this might be a bug.